src/Sets/Zonotope.jl

```diff
@@ -75,40 +90,54 @@ vectors, each vector representing a generator:
 
 ```jldoctest
 julia> Z = Zonotope(ones(2), [[1., 0.], [0., 1.], [1., 1.]])
-Zonotope{Float64}([1.0, 1.0], [1.0 0.0 1.0; 0.0 1.0 1.0])
+Zonotope{Float64,Array{Float64,1},Array{Float64,2}}([1.0, 1.0], [1.0 0.0 1.0; 0.0 1.0 1.0])
 
-julia> Z.generators
+julia> genmat(Z)
 2×3 Array{Float64,2}:
  1.0  0.0  1.0
  0.0  1.0  1.0
 ```
 """
-struct Zonotope{N<:Real} <: AbstractZonotope{N}
-    center::AbstractVector{N}
-    generators::AbstractMatrix{N}
-
-    function Zonotope(center::AbstractVector{N}, generators::AbstractMatrix{N};
-                      remove_zero_generators::Bool=true) where {N<:Real}
+struct Zonotope{N<:Real, VN<:AbstractVector{N}, MN<:AbstractMatrix{N}} <: AbstractZonotope{N}
+    center::VN
+    generators::MN
+
+    function Zonotope(center::VN, generators::MN;
+                      remove_zero_generators::Bool=true) where {N<:Real,
+                                                                VN<:AbstractVector{N},
+                                                                MN<:AbstractMatrix{N}}
         @assert length(center) == size(generators, 1) "the dimension of the " *
             "center ($(length(center))) and the generators " *
             "($(size(generators, 1))) need to match"
         if remove_zero_generators
             generators = delete_zero_columns!(generators)
         end
-        new{N}(center, generators)
+        MT = typeof(generators)
+        new{N, VN, MT}(center, generators)
     end
 end
 
 isoperationtype(::Type{<:Zonotope}) = false
 isconvextype(::Type{<:Zonotope}) = true
 
 # constructor from center and list of generators
-Zonotope(center::AbstractVector{N}, generators_list::AbstractVector{VN};
-         remove_zero_generators::Bool=true
-        ) where {N<:Real, VN<:AbstractVector{N}} =
-    Zonotope(center, hcat(generators_list...);
-             remove_zero_generators=remove_zero_generators)
-
+function Zonotope(center::VN, generators_list::AbstractVector{VN};
+                  remove_zero_generators::Bool=true) where {N<:Real, VN<:AbstractVector{N}}
+    n = length(center)
+    p = length(generators_list)
+    G = Matrix{N}(undef, n, p) # TODO: generalize type, see #2040
+    if remove_zero_generators
+        for (i, gi) in enumerate(generators_list)
+            iszero(gi) && continue
+            @inbounds G[:, i] = gi
+        end
```

Comment:
This does not work for zero generators (there will be `undef` columns). I would just not change the function in this PR, it is unrelated. In any case, you should add a test.

Reply:
okay... i didn't understand what is your use case that doesn't work, but ill defer to a follow up issue :thumbsup:

Reply:
You create a matrix with `p` columns but only insert at column `i` if the vector is not zero. Try calling this function with a single zero vector and you will get an `undef` matrix.